<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Anna Elisabeth Furtjes   King’s College London" />


<title>Genomic PCA</title>

<script src="site_libs/header-attrs-2.10/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/yeti.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Genomic PCA</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="25082021_geneticPCA_explanation.html">Code</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">

<div class="btn-group pull-right float-right">
<button type="button" class="btn btn-default btn-xs btn-secondary btn-sm dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu dropdown-menu-right" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Genomic PCA</h1>
<h4 class="author">Anna Elisabeth Furtjes <br> King’s College London</h4>
<h4 class="date">26/08/2021</h4>

</div>


<p><span class="math display">\[\\[0.1in]\]</span></p>
<hr />
<div id="background-and-reasoning" class="section level1">
<h1>Background and reasoning</h1>
<p><strong>Available data:</strong> In the following analysis, we use freely available genome-wide association study (GWAS) summary statistics from <a href="https://doi.org/10.1038/s41588-018-0309-3">Linner et al. (2019)</a> on risky behaviors . In this study, the authors performed GWAS on the propensity to speeding, alcohol consumption, smoking, and the number of self-reported sexual partners. In addition to the four traits, they extracted the first principal component (PC) of these four traits in a principal component analysis (PCA), and performed a GWAS on this first PC to create a summary of overall risky behaviors. The summary statistics are available on <a href="https://www.thessgac.org/data" class="uri">https://www.thessgac.org/data</a> . The GWAS summary statistics for ever smoking status were derived from UKB as well as TAG, whereas all the others (including the PC sumstats) were obtained from UKB only.</p>
<p><br></p>
<p><strong>Aim of the analysis:</strong> Using Linner’s summary statistics of speeding, drinking, smoking and sexual partners, we will perform linkage disequilibirum score regression (LDSC) and obtain the genetic correlation matrix from the four risky behaviors in Linner et al. (2019). Using PCA, we will extract the first PC from this genetic correlation matrix. Resulting standardised PC loadings will be used to weight SNP effects (z-scores) contributed by the four risky behaviors. We have modified the R package provided by <a href="https://www.ncbi.nlm.nih.gov/pubmed/30643256">Baselman et al. (2019)</a> to create our genetic PC sumstats while allowing to correct for sample overlap.</p>
<p><br></p>
<p><strong>Validation of the resulting genetic PC sumstats:</strong></p>
<p>To validate the results, we will compare SNP effects from Linner’s phenotypic PC GWAS with effect sizes extracted in our analysis. A perfect correlation between our genetic PC sumstats that we created using summary statistics only (called <em>genetic PCA</em> in this document) and the PC sumstats would indicate that we captured the same effects as Linner et al. obtained in their PCA with <em>phenotypic</em> data. Validating the resulting genetic PC sumstats serves as a sanity check and is crucial as, to our knowledge, this kind of approach is unprecedented.</p>
<hr />
<p><span class="math display">\[\\[0.01in]\]</span></p>
</div>
<div id="procedure" class="section level1">
<h1>Procedure</h1>
<div id="a.-obtain-genetic-correlation-matrix" class="section level2">
<h2>A. Obtain genetic correlation matrix</h2>
<p>After downloading the summary statistics, we use the genomic SEM package in R to pre-process and munge the files (<a href="https://www.nature.com/articles/s41562-019-0566-x">Grotzinger et al., 2019</a>). We perform LDSC to obtain a genetic correlation matrix.</p>
<div id="step-1-prepare-and-munge-files" class="section level3">
<h3>Step 1: Prepare and munge files</h3>
<pre class="r"><code>library(data.table)
library(devtools)
library(GenomicSEM)
library(knitr)
library(psych)
# set work directory to where raw GWAS are stored
setwd(paste0(getwd(),&quot;/GWAS_raw&quot;))

# list all files in this directory, only first four are the risky behaviors
gwas_files&lt;-list.files()
gwas_files&lt;-gwas_files[1:4]

#load all files listed in gwas_files
for (i in 1:length(gwas_files)) assign(gwas_files[i],fread(gwas_files[i],header=T,data.table=F))

# double check which files were loaded in 
ls(pattern=&quot;.txt&quot;)

# change column names &amp;
# set wd to the directory the files with the new names should be saved
setwd(paste0(getwd(),&quot;/GWAS_header&quot;))

for(i in gwas_files){
file&lt;-get(i)
print(dim(file))
print(head(file))
names(file)&lt;-c(&quot;SNP&quot;,&quot;CHR&quot;,&quot;POS&quot;,&quot;A1&quot;,&quot;A2&quot;,&quot;eaf_A1&quot;,&quot;beta&quot;,&quot;se&quot;,&quot;p&quot;)
print(head(file))
fwrite(file,file=i, quote=FALSE,col.names=TRUE,row.names=F,sep=&quot; &quot;)
}

#################
## munge
#################
files&lt;-gwas_files

# as reference SNP list use 1000G reference genome (HapMap 3)
# wget https://data.broadinstitute.org/alkesgroup/LDSCORE/w_hm3.snplist.bz2
hm3&lt;-&quot;w_hm3.noMHC.snplist.txt&quot;

trait.names&lt;-c(&quot;speeding&quot;,&quot;drinking&quot;,&quot;smoking&quot;,&quot;sexual_partners&quot;)

N&lt;-c(404291,414343,518633,370711)

munge(files=files,
      hm3=hm3,
      trait.names = trait.names,
      N=N)</code></pre>
</div>
<div id="step-2-perform-linkage-disequilibrium-score-regression-ldsc" class="section level3">
<h3>Step 2: Perform linkage disequilibrium score regression (LDSC)</h3>
<pre class="r"><code># download and unpack SNP weights: 
# wget https://data.broadinstitute.org/alkesgroup/LDSCORE/eur_w_ld_chr.tar.bz2
# tar -jxvf eur_w_ld_chr.tar.bz2
# download software to unzip .tar document on windows
##############################################################
# setwd to munged files
setwd(paste0(getwd(),&quot;/GWAS_munged&quot;))

library(stringr)

# specifications for ldsc function
traits&lt;-c(&quot;speeding.sumstats.gz&quot;,&quot;drinking.sumstats.gz&quot;,&quot;smoking.sumstats.gz&quot;,&quot;sexual_partners.sumstats.gz&quot;)
ld&lt;- &quot;eur_w_ld_chr/&quot;
wld&lt;- &quot;eur_w_ld_chr/&quot;
trait.names&lt;-str_remove(traits, pattern = &quot;.sumstats.gz&quot;)
sample.prev&lt;-c(NA,NA,.59,NA)

# https://bmjopen.bmj.com/content/4/12/e005663 
# this paper says that 50% of population had never smoked (but they use UKB ...?)
# difficult to find papers on ever smoked rather than regular smokers
population.prev&lt;-c(NA,NA,.5,NA)

LDSCoutput_risk_taking&lt;-ldsc(traits=traits,
                             sample.prev=sample.prev,
                             population.prev=population.prev,
                             ld=ld,wld=wld,
                             trait.names = trait.names,
                             stand = T)

#LDSCoutput_risk_taking$S is going to be the covariance matrix
#LDSCoutput_risk_taking$V is the sampling covariance matrix as expected by Lavaan
save(LDSCoutput_risk_taking, file=&quot;risk_taking_stand.RData&quot;)</code></pre>
</div>
<div id="step-3-display-genetic-correlation-matrix" class="section level3">
<h3>Step 3: Display genetic correlation matrix</h3>
<p><img src="25082021_geneticPCA_explanation_files/figure-html/heatmap-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="sanity-check" class="section level3">
<h3><strong>Sanity check:</strong></h3>
<p>To validate our correlation matrix, we contacted Linner et al. (2019) who shared their LDSC results: They found very similar genetic correlations between the four risky behaviors using LDSC and we suggest that our pre-processing was successful.</p>
<p><img src="25082021_geneticPCA_explanation_files/figure-html/Linners_rg-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="ldsc-intercepts" class="section level3">
<h3>LDSC intercepts</h3>
<p>LDSC intercepts quantify sample overlap and other confounding factors (such as phenotypic correlation and population stratification) between the four risky behaviors we are analysing here.</p>
<p><img src="25082021_geneticPCA_explanation_files/figure-html/heatmap_intercept_four_traits-1.png" width="672" style="display: block; margin: auto;" /></p>
<p><br></p>
<hr />
</div>
</div>
<div id="b.-principal-component-analysis" class="section level2">
<h2>B. Principal Component Analysis</h2>
<div id="eigen-decomposition" class="section level3">
<h3>Eigen decomposition</h3>
<p>We perform eigen decomposition on the <strong>correlation matrix</strong> using the <code>eigen</code> function. We aim to obtain standardised loadings of the four risky behaviors on the first PC, by standardising the eigenvectors of the first PC with their corresponding eigenvalue.</p>
</div>
<div id="eigenvectors" class="section level3">
<h3>Eigenvectors</h3>
<pre class="r"><code>#extract eigenvectors and eigenvalues
eigenvectors&lt;-round(eigen(cormatrix)$vectors,digits=3)</code></pre>
<p><span class="math display">\[\begin{bmatrix}
0.37&amp;0.853&amp;0.249&amp;-0.27 \\
0.498&amp;0.094&amp;-0.837&amp;0.206 \\
0.544&amp;-0.481&amp;0.102&amp;-0.68 \\
0.565&amp;-0.179&amp;0.476&amp;0.65 \\
\end{bmatrix}\]</span></p>
</div>
<div id="eigenvalues-and-explained-variance" class="section level3">
<h3>Eigenvalues and explained variance</h3>
<pre class="r"><code>eigenvalues&lt;-round(eigen(cormatrix)$values,digits = 3)
explained_variance&lt;-round(eigenvalues/sum(eigenvalues)*100,digits = 3)
library(knitr)
values&lt;-cbind(c(&quot;1st PC&quot;,&quot;2nd PC&quot;,&quot;3rd PC&quot;,&quot;4th PC&quot;),eigenvalues,explained_variance)
kable(values, caption=&quot;Eigenvalues and explained variance (%)&quot;, col.names = c(&quot;PC&quot;,&quot;Eigenvalues&quot;,&quot;Explained variance&quot;))</code></pre>
<table>
<caption>Eigenvalues and explained variance (%)</caption>
<thead>
<tr class="header">
<th align="left">PC</th>
<th align="left">Eigenvalues</th>
<th align="left">Explained variance</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">1st PC</td>
<td align="left">1.996</td>
<td align="left">49.888</td>
</tr>
<tr class="even">
<td align="left">2nd PC</td>
<td align="left">0.897</td>
<td align="left">22.419</td>
</tr>
<tr class="odd">
<td align="left">3rd PC</td>
<td align="left">0.685</td>
<td align="left">17.121</td>
</tr>
<tr class="even">
<td align="left">4th PC</td>
<td align="left">0.423</td>
<td align="left">10.572</td>
</tr>
</tbody>
</table>
<p><br></p>
</div>
<div id="standardised-pc-loadings" class="section level3">
<h3>Standardised PC loadings</h3>
<p>We standardise the eigenvectors with the eigenvalues, in order to obtain standardised PC loadings.</p>
<p><img src="standardise_loadings.PNG" style="width:40.0%" /></p>
<p><br></p>
<p>The PC loadings from our genetic <strong>correlation</strong> matrix are: <span class="math display">\[\begin{bmatrix}
0.523&amp;0.808&amp;0.206&amp;-0.176 \\
0.704&amp;0.089&amp;-0.693&amp;0.134 \\
0.769&amp;-0.456&amp;0.084&amp;-0.442 \\
0.798&amp;-0.17&amp;0.394&amp;0.423 \\
\end{bmatrix}\]</span></p>
<p><br> <br></p>
<hr />
</div>
</div>
<div id="c.-create-genetic-pc-sumstats" class="section level2">
<h2>C. Create genetic PC sumstats</h2>
<p>We format the files as expected by the <code>multivariate_GWAMA()</code> function by Baselman et al. (2019). It expects exactly nine columns (SNPID,CHR,BP,EA,OA,EAF,N,Z,P) and that effect sizes must be z scores. We modified the function from Baselman et al. as displayed below in order to create our genetic PC sumstats.</p>
<p><br></p>
<div id="step-1-format-the-risky-behaviors-sumstats" class="section level3">
<h3>Step 1: Format the risky behaviors sumstats</h3>
<pre class="r"><code>source(&quot;https://github.com/baselmans/multivariate_GWAMA/blob/master/Test_Data/N_weighted_GWAMA.function.1_2_6.R?raw=TRUE&quot;)
library(data.table)
setwd(paste0(getwd(),&quot;/GWAS_header&quot;))

# read in gwas files
gwas_files&lt;-list.files()
gwas_files&lt;-gwas_files[1:4]

for (i in 1:length(gwas_files)) assign(gwas_files[i],fread(gwas_files[i],header=T,data.table=F))
ls(pattern=&quot;.txt&quot;)

# generate N column
AUTOMOBILE_SPEEDING_PROPENSITY_GWAS.txt$N&lt;-404291
DRINKS_PER_WEEK_GWAS.txt$N&lt;-414343
EVER_SMOKER_GWAS_MA_UKB_TAG.txt$N&lt;-518633
NUMBER_SEXUAL_PARTNERS_GWAS.txt$N&lt;-370711
  
# the following loop does:
# restrict file to HapMap SNPs contained in munged GWAS (but make sure EAF taken from reference file is correct by comparing EA in ref and munged files)
# transform beta into z-scores
# z-score=beta/stand_error # https://www.biostars.org/p/140292/
# re-name columns and include only the 9 columns needed by the function

# list reference files that contain info on BP, CHR and P (that the munged files don&#39;t contain)
setwd(paste0(getwd(),&quot;/GWAS_header&quot;))
ref_files&lt;-list.files()[1:4]

all_files &lt;- data.frame(ref_files, munged_files = c(&quot;speeding.sumstats.gz&quot;,&quot;drinking.sumstats.gz&quot;,&quot;smoking.sumstats.gz&quot;,&quot;sexual_partners.sumstats.gz&quot;))


for(i in 1:nrow(all_files)){
  # read in munged file
  setwd(paste0(getwd(),&quot;/GWAS_munged&quot;))
  munged_file&lt;-fread(as.character(all_files$munged_files[i]),header=T,data.table=F)
  
  # read in ref file
  setwd(paste0(getwd(),&quot;/GWAS_header&quot;))
  ref_file&lt;-fread(as.character(all_files$ref_files[i]),header=T,data.table=F)
  
  # format ref file
    names(ref_file)[which(names(ref_file)==&quot;POS&quot;)]&lt;-&quot;BP&quot;
    names(ref_file)[which(names(ref_file)==&quot;eaf_A1&quot;)]&lt;-&quot;EAF_A1_prelim&quot;
    names(ref_file)[which(names(ref_file)==&quot;p&quot;)]&lt;-&quot;P&quot;
    # A1 is the effect allele according to Linner website
    names(ref_file)[which(names(ref_file)==&quot;A1&quot;)]&lt;-&quot;EA&quot;
    names(ref_file)[which(names(ref_file)==&quot;A2&quot;)]&lt;-&quot;OA&quot;

    
  # only keep relevant columns from the reference file
  ref_file&lt;-ref_file[,c(&quot;BP&quot;,&quot;CHR&quot;,&quot;SNP&quot;,&quot;P&quot;,&quot;EA&quot;,&quot;EAF_A1_prelim&quot;)]
  
  # merge munged and ref file
  merged_file&lt;-merge(munged_file, ref_file, by=c(&quot;SNP&quot;), all.x=T)
  
  # create an effect allele frequency column that matches the allele used in the munged file (because we take the frequency from the reference file that might have flipped effect and other alleles)
    merged_file$EAF&lt;-ifelse(merged_file$A1 == merged_file$EA, merged_file$EAF_A1_prelim, 1-merged_file$EAF_A1_prelim)
    
    # only keep required columns
    names(merged_file)[which(names(merged_file) == &quot;SNP&quot;)]&lt;-&quot;SNPID&quot;
    merged_file&lt;-merged_file[,c(&quot;SNPID&quot;,&quot;CHR&quot;,&quot;BP&quot;,&quot;A1&quot;,&quot;A2&quot;,&quot;EAF&quot;,&quot;N&quot;,&quot;Z&quot;,&quot;P&quot;)]
    
    # name according to GWAMA expectations
    # In genomicSEM A1 is the effect allele
    names(merged_file)[which(names(merged_file) == &quot;A1&quot;)]&lt;-&quot;EA&quot;
    names(merged_file)[which(names(merged_file) == &quot;A2&quot;)]&lt;-&quot;OA&quot;
    
    # save formatted files
    setwd(paste0(getwd(),&quot;/GWAMA_format&quot;))

    ref_trait &lt;- as.character(all_files$ref_files[i])
    fwrite(merged_file,file=paste0(&quot;gwama_&quot;,ref_trait), quote=FALSE,col.names=TRUE,row.names=F,sep=&quot; &quot;,na=&quot;NA&quot;)
}</code></pre>
<p><br></p>
</div>
<div id="step-2-modify-gwama-function" class="section level3">
<h3>Step 2: Modify GWAMA function</h3>
<p><img src="Baselman_equation.PNG" style="width:45.0%" /></p>
<p>Baselman et al. (2019) implemeted the equation above. We replace the weights in their equation, the square root of N times the heritability, with standardised PC loadings times the square root of N:</p>
<p>Change <code>w = sqrt(N*h2)</code>to <code>w = sqrt(N)*standardised_loadings</code></p>
<p>If we only replace heritabilities with standardised loadings in the arguments of the function, we will get errors for two reasons:</p>
<ol style="list-style-type: decimal">
<li><p>It won’t allow negative loadings (because we are replacing h2 with potentially negative PC loadings and negative heritability wouldn’t make sense)</p></li>
<li><p>It won’t allow to calculate our weights, as there is no sqrt() of a negative value: <code>sqrt(-number)=NaN</code></p></li>
</ol>
<p>I have made the following changes to address these problems:</p>
<div id="disable-sanity-check-for-negative-heritabilites" class="section level4">
<h4>1) Disable sanity check for negative heritabilites:</h4>
<pre class="r"><code>source(&quot;https://github.com/baselmans/multivariate_GWAMA/blob/master/Test_Data/N_weighted_GWAMA.function.1_2_6.R?raw=TRUE&quot;)
my_GWAMA&lt;-multivariate_GWAMA
trace(my_GWAMA,edit=T)</code></pre>
<p>I deleted <em>lines 217-221</em>, which disables the sanity check to stop the function from running when loadings are not in the expected h2 format (h2&lt;0 &amp; h2&gt;1).</p>
</div>
<div id="modify-code-to-calculate-our-weights" class="section level4">
<h4>2) Modify code to calculate our weights</h4>
<p><strong>Original code:</strong></p>
<pre class="r"><code># line 663
W&lt;-t(t(N) * h2)
# line 665
sqrt_W&lt;-sqrt(W)</code></pre>
<p><em>As we replace h2 with negative loadings, this code produces NaNs.</em></p>
<p><strong>Modified code:</strong></p>
<pre class="r"><code>w = t(t(sqrt(N))*h2)
sqrt_W&lt;-w</code></pre>
<p><br></p>
</div>
</div>
<div id="step-3-run-modified-function" class="section level3">
<h3>Step 3: Run modified function</h3>
<pre class="r"><code># load dependencies
source(&quot;https://github.com/baselmans/multivariate_GWAMA/blob/master/Test_Data/N_weighted_GWAMA.function.1_2_6.R?raw=TRUE&quot;)
library(data.table)
library(stringr)

setwd(paste0(substring(getwd(), first = 1, last = 18),&quot;Documents/GitHub/genomicPCA/scripts&quot;))
source(&quot;my_GWAMA_26032020.R&quot;)


# Make an empty list with length equal to the number of input files
dat&lt;-vector(&quot;list&quot;)

setwd(paste0(getwd(),&quot;/GWAMA_format&quot;))
gwama_files&lt;-list.files(pattern = &quot;gwama_&quot;)
all_files&lt;-as.data.frame(cbind(gwama_files, file_names = c(&quot;speeding&quot;,&quot;drinking&quot;,&quot;smoking&quot;,&quot;sexual_partners&quot;)))
  
# Read in the data 
n&lt;-0
for(file in 1:nrow(all_files)){
  n&lt;-n+1
  print(n)
  dat[[all_files$file_names[n]]]&lt;-fread(all_files$gwama_files[n],data.table = F)
}

# load covariance matrix from LDSC
setwd(paste0(substring(getwd(), first = 1, last = 18),&quot;Documents/GitHub/genomicPCA/data&quot;))
load(&quot;risk_taking_stand.RData&quot;)

# select the matrix of LDSC intercepts 
# this matrix should be in the same order as data read into dat but I have added some code to specifically make sure they take the same order
order&lt;-names(dat)
dimnames(LDSCoutput_risk_taking$I)[[1]]&lt;-dimnames(LDSCoutput_risk_taking$S)[[2]]
dimnames(LDSCoutput_risk_taking$I)[[2]]&lt;-dimnames(LDSCoutput_risk_taking$S)[[2]]
CTI&lt;-as.matrix(LDSCoutput_risk_taking$I[order,order])

# load correlation matrix
dimnames(LDSCoutput_risk_taking$S_Stand)[[1]]&lt;-dimnames(LDSCoutput_risk_taking$S)[[2]]
dimnames(LDSCoutput_risk_taking$S_Stand)[[2]]&lt;-dimnames(LDSCoutput_risk_taking$S)[[2]]
cormatrix&lt;-LDSCoutput_risk_taking$S_Stand[order,order]

# eigen decomposition
eigenvectors&lt;-eigen(cormatrix)$vectors
eigenvalues&lt;-eigen(cormatrix)$values

# calculate standardised loadings
loadings&lt;-as.vector(eigenvectors%*%sqrt(diag(eigenvalues))[,1])

# run my modified GWAMA function
setwd(paste0(getwd(),&quot;/Evaluate_GWAMA&quot;))

my_GWAMA(x=dat,
          cov_Z=CTI,
          h2=loadings,
          out=&quot;.&quot;,
          name=&quot;my_GWAMA_weights_26082021&quot;,
          output_gz=F,
          check_columns=F)</code></pre>
<p>Running the function can take a while. It will create one file containing the genetic PC summary stats and a log file to keep track of operations.</p>
<p>For example, the log file shows the number of total and significant SNPs:</p>
<ol style="list-style-type: decimal">
<li><p>“2,529 SNPs reached genome-wide significance (Minimum P-value: 9.4467e-41)”</p></li>
<li><p>“Successfully created summary statistics for 1,198,974 SNPs”</p></li>
</ol>
<p><em>The function does not test for significance. It only counts the number of SNP’s below a certain p-value threshold. The default threshold used here is <code>p &lt; 5e-08</code>, but can be changed by the user.</em></p>
<p><br></p>
<p><br></p>
<hr />
</div>
</div>
</div>
<div id="evaluate-results" class="section level1">
<h1>Evaluate results</h1>
<p>We chose to evaluate our genetic PC summary statistics using data from Linner et al. (2019) because they provide GWAS on a first phenotypic PC derived from the four GWAS phenotypes. We use their phenotypic PC sumstats as a reference for how well our genetic PC represents a summary of SNP effects contributed by the four risky behaviors: speeding, drinking, smoking and sexual partners. If our genetic PC summary statistics highly correlate with the phenotypic PC sumstats by Linner, we can be confident that our genetic PC extracted purely based on GWAS sumstats is capturing the same SNP effects as GWAS sumstats calculated for a phenotypic PC.</p>
<p>We will perform genetic correlations between our genetic PC sumstats, the phenotypic PC sumstats and the four risky behaviors summary stats. We will also plot z statistics contained in our genetic PC sumstats against those contained in the other summary statistics and calculate their correlations.</p>
<p>It is important to note that the directions indicated by the PC loadings are arbitrary and it would not change the signal contained in the statistics if we flipped the signs of the loadings as long as this is done consistently across all PC loadings. If negative PC loadings are obtained, one can weight the SNPs with negative PC loadings, but we suggest to scale the SNP effects in the most sensible direction to make results more intuitively interpretable.</p>
<div id="prepare-and-munge-sumstats-for-ldsc" class="section level3">
<h3>Prepare and munge sumstats for LDSC</h3>
<pre class="r"><code># set work directory to where raw GWAS are stored
setwd(paste0(getwd(),&quot;/GWAS_raw&quot;))

# store file name in gwas_file
gwas_file&lt;-list.files(pattern=&quot;RISK_PC&quot;)

# load into R
Risk_PC&lt;-fread(gwas_file,header=T,data.table = F)

# rename columns
names(Risk_PC)&lt;-c(&quot;SNP&quot;,&quot;CHR&quot;,&quot;POS&quot;,&quot;A1&quot;,&quot;A2&quot;,&quot;eaf_A1&quot;,&quot;beta&quot;,&quot;se&quot;,&quot;p&quot;)

# and set wd to the directory I want to save the files with the new names
setwd(paste0(getwd(),&quot;/Evaluate_GWAMA&quot;))

# write table 
fwrite(Risk_PC,file=&quot;Risk_PC_GWAS.txt&quot;, quote=FALSE,col.names=TRUE,row.names=F,sep=&quot; &quot;)


# prepare GWAMA file #
# set work directory to where GWAMA file is stored
setwd(paste0(getwd(),&quot;/Evaluate_GWAMA&quot;))
# store file name in gwas_file
gwas_file&lt;-list.files(pattern=&quot;weighted_GWAMA.results.txt&quot;)
# load into R
GWAMA&lt;-fread(gwas_file,header=T,data.table = F)
# rename columns (rename N_eff into N so that GenomicSEM recognises it as N column)
names(GWAMA)[which(names(GWAMA)==&quot;N_eff&quot;)]&lt;-&quot;N&quot;
GWAMA&lt;-GWAMA[,c(&quot;SNPID&quot;,&quot;CHR&quot;,&quot;BP&quot;,&quot;EA&quot;,&quot;OA&quot;,&quot;EAF&quot;,&quot;MAF&quot;,&quot;N&quot;,&quot;BETA&quot;,&quot;SE&quot;,&quot;Z&quot;,&quot;p&quot;)]

# setwd to where you want to save GWAMA
setwd(paste0(getwd(),&quot;/Evaluate_GWAMA&quot;))
# write table 
fwrite(GWAMA,file=&quot;GWAMA_header.txt&quot;, quote=FALSE,col.names=TRUE,row.names=F,sep=&quot; &quot;,na=&quot;NA&quot;)

# munge PC and GWAMA files ##

#load dependencies for genomic SEM
library(devtools)
library(GenomicSEM)

files&lt;-c(&quot;Risk_PC_GWAS.txt&quot;,&quot;GWAMA_header.txt&quot;)

# use 1000G reference genome (HapMap 3)
# wget https://data.broadinstitute.org/alkesgroup/LDSCORE/w_hm3.snplist.bz2
hm3&lt;-paste0(getwd(),&quot;/GWAS_header/w_hm3.noMHC.snplist.txt&quot;)

trait.names&lt;-c(&quot;Risk_PC&quot;,&quot;GWAMA&quot;)

N&lt;-c(315894,NA)

munge(files=files,
      hm3=hm3,
      trait.names = trait.names,
      N=N)</code></pre>
</div>
<div id="perform-ldsc" class="section level3">
<h3>Perform LDSC</h3>
<pre class="r"><code>setwd(paste0(getwd(),&quot;/GWAS_munged&quot;))

traits&lt;-c(&quot;speeding.sumstats.gz&quot;,&quot;drinking.sumstats.gz&quot;,&quot;smoking.sumstats.gz&quot;,&quot;sexual_partners.sumstats.gz&quot;,paste0(getwd(),&quot;/Evaluate_GWAMA/Risk_PC.sumstats.gz&quot;),paste0(getwd(),&quot;/Evaluate_GWAMA/GenomicPCA.sumstats.gz&quot;))

ld&lt;- &quot;eur_w_ld_chr/&quot;
wld&lt;- &quot;eur_w_ld_chr/&quot;
trait.names&lt;-c(&quot;speeding&quot;,&quot;drinking&quot;,&quot;smoking&quot;,&quot;sexual_partners&quot;,&quot;phenotypicPC&quot;,&quot;genomicPC&quot;)
sample.prev&lt;-c(NA,NA,.59,NA,NA,NA)
population.prev&lt;-c(NA,NA,.5,NA,NA,NA)

LDSCoutput_evaluate_GWAMA&lt;-ldsc(traits=traits,
                                 sample.prev=sample.prev,
                                 population.prev=population.prev,
                                 ld=ld,wld=wld,
                                 trait.names = trait.names, stand = T)

setwd(paste0(getwd(),&quot;/Evaluate_GWAMA&quot;))
save(LDSCoutput_evaluate_GWAMA, file=&quot;Evaluate_GWAMA.RData&quot;)</code></pre>
</div>
<div id="display-genetic-correlations" class="section level3">
<h3>Display genetic correlations</h3>
<p><img src="25082021_geneticPCA_explanation_files/figure-html/heatmap_GWAMA-1.png" width="672" style="display: block; margin: auto;" /></p>
<p><em>Note the strong genetic correlation between our genomic PCA sumstats and the phenotypic PC sumstats. It suggests that we extract the same SNP effects from GWAS summary statistics only as was captured by Linner when extracting a phenotypic PC and calculating GWAS sumstats for it.</em></p>
</div>
<div id="display-ldsc-intercepts" class="section level3">
<h3>Display LDSC intercepts</h3>
<p>LDSC intercepts capture sample overlap, among other confounding factors. We would expect traits compared with themselves to have a perfect sample overlap (1 or very close to 1). When comparing our genetic PC sumstats with the four risky behaviors sumstats, we expect substantial sample overlap as all four have contributed to our genetic PC. In theory, our genetic PC and the phenotypic PC by Linner capture the same signal, we expect an even stronger sample overlap between these two traits.</p>
<pre class="r"><code>setwd(paste0(substring(getwd(), first = 1, last = 18),&quot;Documents/GitHub/genomicPCA/data&quot;))
load(&quot;Evaluate_GWAMA.RData&quot;)

dimnames(LDSCoutput_evaluate_GWAMA$I)[[1]]&lt;-dimnames(LDSCoutput_evaluate_GWAMA$S)[[2]]
dimnames(LDSCoutput_evaluate_GWAMA$I)[[2]]&lt;-dimnames(LDSCoutput_evaluate_GWAMA$S)[[2]]

intercepts&lt;-data.matrix(round(LDSCoutput_evaluate_GWAMA$I,digits = 2))</code></pre>
<p><img src="25082021_geneticPCA_explanation_files/figure-html/heatmap_intercepts-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="comparison-of-our-genetic-pc-sumstats-with-other-traits" class="section level3">
<h3>Comparison of our genetic PC sumstats with other traits</h3>
<p>In the following plots, the blue line indicates the central correlation line and the red line is the line of identity with an intercept of 1 and a slope of -1: <code>x=y</code>.</p>
<div id="our-genetic-pc-sumstats-vs.-linners-phenotypic-pc-sumstats" class="section level4">
<h4>Our genetic PC sumstats vs. Linner’s phenotypic PC sumstats</h4>
<p><img src="plots/pheno_vs_genoPC.png" /></p>
</div>
<div id="our-genetic-pc-sumstats-vs.-speeding-sumstats" class="section level4">
<h4>Our genetic PC sumstats vs. speeding sumstats</h4>
<p><img src="plots/speeding_vs_genoPC.png" /></p>
</div>
<div id="our-genetic-pc-sumstats-vs.-drinking-sumstats" class="section level4">
<h4>Our genetic PC sumstats vs. drinking sumstats</h4>
<p><img src="plots/drinking_vs_genoPC.png" /></p>
</div>
<div id="our-genetic-pc-sumstats-vs.-smoking-sumstats" class="section level4">
<h4>Our genetic PC sumstats vs. smoking sumstats</h4>
<p><img src="plots/smoking_vs_genoPC.png" /></p>
</div>
<div id="our-genetic-pc-sumstats-vs.-sexual-partners-sumstats" class="section level4">
<h4>Our genetic PC sumstats vs. sexual partners sumstats</h4>
<p><img src="plots/sexual_vs_genoPC.png" /></p>
<p><br></p>
<p><br></p>
<hr />
</div>
</div>
</div>
<div id="alternative-procedure" class="section level1">
<h1>Alternative procedure</h1>
<p>For comparison, we also calculated genetic PC sumstats using a covariance matrix obtained through LDSC, instead of a correlation matrix. We run the same code as displayed above (Step 3: Run GWAMA with modified function), but replacing the standardised genetic correlation matrix with the unstandardised matrix.</p>
<div id="genetic-pc-sumstats-extracted-from-covariance-matrix" class="section level3">
<h3>Genetic PC sumstats extracted from covariance matrix</h3>
<pre class="r"><code># load dependencies
source(&quot;https://github.com/baselmans/multivariate_GWAMA/blob/master/Test_Data/N_weighted_GWAMA.function.1_2_6.R?raw=TRUE&quot;)
library(data.table)
library(stringr)

setwd(paste0(substring(getwd(), first = 1, last = 18),&quot;Documents/GitHub/genomicPCA/scripts&quot;))
source(&quot;my_GWAMA_26032020.R&quot;)


# Make an empty list with length equal to the number of input files
dat&lt;-vector(&quot;list&quot;)


setwd(paste0(getwd(),&quot;/GWAMA_format&quot;))
gwama_files&lt;-list.files(pattern = &quot;gwama_&quot;)
all_files&lt;-as.data.frame(cbind(gwama_files, file_names = c(&quot;speeding&quot;,&quot;drinking&quot;,&quot;smoking&quot;,&quot;sexual_partners&quot;)))
  
# Read in the data 
n&lt;-0
for(file in 1:nrow(all_files)){
  n&lt;-n+1
  print(n)
  dat[[all_files$file_names[n]]]&lt;-fread(all_files$gwama_files[n],data.table = F)
}

# load covariance matrix from LDSC
setwd(paste0(substring(getwd(), first = 1, last = 18),&quot;Documents/GitHub/genomicPCA/data&quot;))
load(&quot;risk_taking_stand.RData&quot;)

# select the matrix of LDSC intercepts 
# this matrix should be in the same order as data read into dat but I have added some code to specifically make sure they take the same order
order&lt;-names(dat)
dimnames(LDSCoutput_risk_taking$I)[[1]]&lt;-dimnames(LDSCoutput_risk_taking$S)[[2]]
dimnames(LDSCoutput_risk_taking$I)[[2]]&lt;-dimnames(LDSCoutput_risk_taking$S)[[2]]
CTI&lt;-as.matrix(LDSCoutput_risk_taking$I[order,order])

# load correlation matrix
dimnames(LDSCoutput_risk_taking$S)[[1]]&lt;-dimnames(LDSCoutput_risk_taking$S)[[2]]
covmatrix&lt;-LDSCoutput_risk_taking$S[order,order]

# eigen decomposition
eigenvectors&lt;-eigen(covmatrix)$vectors
eigenvalues&lt;-eigen(covmatrix)$values

# calculate standardised loadings
loadings&lt;-as.vector(eigenvectors%*%sqrt(diag(eigenvalues))[,1])

    # multiply the loadings with -1 if the median value of the loadings is below 0
    # this achieves a more intuitive interpretation but doesn&#39;t change the results - the signal is the same even when we multiple it with -1
    median_loadings&lt;-median(loadings)
    mean_direction&lt;-sign(median_loadings)
    if(mean_direction == -1){
    loadings&lt;-loadings*(-1)
    } else {
    loadings&lt;-loadings}

# run my modified GWAMA function
setwd(paste0(getwd(),&quot;/Evaluate_GWAMA&quot;))
my_GWAMA(x=dat,
          cov_Z=CTI,
          h2=loadings,
          out=&quot;.&quot;,
          name=&quot;my_GWAMA_covmatrix_26082021&quot;,
          output_gz=F,
          check_columns=F)</code></pre>
<div id="munge-and-format-files" class="section level4">
<h4>Munge and format files</h4>
<pre class="r"><code># prepare GWAMA file #
# set work directory to where GWAMA file is stored
setwd(paste0(getwd(),&quot;/Evaluate_GWAMA&quot;))
# store file name in object called file
file&lt;-list.files(pattern=&quot;covmatrix_26082021.N_weighted_GWAMA.results&quot;)
# load into R
library(data.table)
GWAMA&lt;-fread(file,header=T,data.table = F)

# rename columns (rename N_eff into N so that GenomicSEM recognises it as N column)
names(GWAMA)[which(names(GWAMA)==&quot;N_eff&quot;)]&lt;-&quot;N&quot;
names(GWAMA)[which(names(GWAMA)==&quot;PVAL&quot;)]&lt;-&quot;P&quot;

GWAMA&lt;-GWAMA[,c(&quot;SNPID&quot;,&quot;CHR&quot;,&quot;BP&quot;,&quot;EA&quot;,&quot;OA&quot;,&quot;EAF&quot;,&quot;MAF&quot;,&quot;N&quot;,&quot;BETA&quot;,&quot;SE&quot;,&quot;Z&quot;,&quot;P&quot;)]

# setwd to where you want to save GWAMA
setwd(paste0(getwd(),&quot;/Evaluate_GWAMA&quot;))
# write table 
fwrite(GWAMA,file=&quot;GWAMA_cov_header.txt&quot;, quote=FALSE,col.names=TRUE,row.names=F,sep=&quot; &quot;,na=&quot;NA&quot;)

#load dependencies for genomic SEM
library(devtools)
library(GenomicSEM)

files&lt;-&quot;GWAMA_cov_header.txt&quot;

# use 1000G reference genome (HapMap 3)
# wget https://data.broadinstitute.org/alkesgroup/LDSCORE/w_hm3.snplist.bz2
hm3&lt;-paste0(getwd(),&quot;/GWAS_header/w_hm3.noMHC.snplist.txt&quot;)

trait.names&lt;-&quot;GenomicPCA_cov&quot;

N&lt;-NA

munge(files=files,
      hm3=hm3,
      trait.names = trait.names,
      N=N)</code></pre>
</div>
<div id="calculate-ldsc" class="section level4">
<h4>Calculate LDSC</h4>
<pre class="r"><code>setwd(paste0(getwd(),&quot;/GWAS_munged&quot;))

traits&lt;-c(&quot;speeding.sumstats.gz&quot;,&quot;drinking.sumstats.gz&quot;,&quot;smoking.sumstats.gz&quot;,&quot;sexual_partners.sumstats.gz&quot;,paste0(getwd(),&quot;/Evaluate_GWAMA/Risk_PC.sumstats.gz&quot;),paste0(getwd(),&quot;/Evaluate_GWAMA/GenomicPCA_cov.sumstats.gz&quot;))

ld&lt;- &quot;eur_w_ld_chr/&quot;
wld&lt;- &quot;eur_w_ld_chr/&quot;
trait.names&lt;-c(&quot;speeding&quot;,&quot;drinking&quot;,&quot;smoking&quot;,&quot;sexual_partners&quot;,&quot;phenotypicPC&quot;,&quot;genomicPC&quot;)
sample.prev&lt;-c(NA,NA,.59,NA,NA,NA)
population.prev&lt;-c(NA,NA,.5,NA,NA,NA)

LDSCoutput_evaluate_GWAMA&lt;-ldsc(traits=traits,
                                 sample.prev=sample.prev,
                                 population.prev=population.prev,
                                 ld=ld,wld=wld,
                                 trait.names = trait.names, stand = T)

setwd(paste0(getwd(),&quot;/Evaluate_GWAMA&quot;))
save(LDSCoutput_evaluate_GWAMA, file=&quot;Evaluate_cov_GWAMA.RData&quot;)</code></pre>
</div>
<div id="heatmap" class="section level4">
<h4>Heatmap</h4>
<p><img src="25082021_geneticPCA_explanation_files/figure-html/heatmap_covariance-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>The resulting genetic PC sumstats weighted by PC loadings extracted from a covariance matrix index very similar signal as those extracted from a correlation matrix.</p>
<p>By using a covariance instead of a correlation matrix, the genetic correlation between our genotypic PC sumstats and the phenotypic PC sumstats slighly declined (rcor = 0.99 vs. rcov = -0.95), suggesting that using PC loadings from a correlation matrix to weigh the SNP-weights produced slightly more accurate results. The genetic correlation between our genotypic PC and individual risky behaviors also slightly shifted.</p>
<p><br> <br></p>
<hr />
</div>
</div>
<div id="compare-our-genetic-pc-sumstats-to-other-traits" class="section level3">
<h3>Compare our genetic PC sumstats to other traits</h3>
<p>These genetic PC sumstats have been extracted from a covariance instead of a correlation matrix. In the following plots, the blue line indicates the central correlation line and the red line is the line of identity with an intercept of 1 and a slope of -1: <code>x=-y</code>.</p>
<div id="our-genetic-pc-sumstats-vs.-linners-phenotypic-pc-sumstats-1" class="section level4">
<h4>Our genetic PC sumstats vs. Linner’s phenotypic PC sumstats</h4>
<p><img src="plots/pheno_vs_genoPC_unstand.png" /></p>
</div>
<div id="our-genetic-pc-sumstats-vs.-speeding-sumstats-1" class="section level4">
<h4>Our genetic PC sumstats vs. speeding sumstats</h4>
<p><img src="plots/speeding_vs_genoPC_unstand.png" /></p>
</div>
<div id="our-genetic-pc-sumstats-vs.-drinking-sumstats-1" class="section level4">
<h4>Our genetic PC sumstats vs. drinking sumstats</h4>
<p><img src="plots/drinking_vs_genoPC_unstand.png" /></p>
</div>
<div id="our-genetic-pc-sumstats-vs.-smoking-sumstats-1" class="section level4">
<h4>Our genetic PC sumstats vs. smoking sumstats</h4>
<p><img src="plots/smoking_vs_genoPC_unstand.png" /></p>
</div>
<div id="our-genetic-pc-sumstats-vs.-sexual-partners-sumstats-1" class="section level4">
<h4>Our genetic PC sumstats vs. sexual partners sumstats</h4>
<p><img src="plots/sexual_vs_genoPC_unstand.png" /></p>
</div>
</div>
</div>

&nbsp;
<hr />
<a href = "https://www.kcl.ac.uk/">
<p style="text-align: center"><img src="KCL_logo.jpg" href="https://www.kcl.ac.uk/" style="width:70px;height:50px"> 
</a>
</p>

<a href = "https://www.kcl.ac.uk/people/anna-furtjes">
<p style="text-align: center;">By Anna Elisabeth Fürtjes</a></p>
<p style="text-align: center;"><span style="color: #808080;"><em>anna.furtjes@kcl.ac.uk</em></span></p>

<!-- Add icon library -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">

<!-- Add font awesome icons -->
<p style="text-align: center;">
    <a href="https://twitter.com/Anna_Furtjes" class="fa fa-twitter"></a>
</p>

&nbsp;


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("show" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
